internal/controller/ngrok/kubernetesoperator_controller.go

```diff
@@ -0,0 +1,453 @@
+	"github.com/ngrok/ngrok-api-go/v6"
+	ngrokv1alpha1 "github.com/ngrok/ngrok-operator/api/ngrok/v1alpha1"
+	"github.com/ngrok/ngrok-operator/internal/controller"
+	"github.com/ngrok/ngrok-operator/internal/ngrokapi"
+)
+
+// TODO: features need to be capitalized in the ngrok API currently, this is subject to change
+var featureMap = map[string]string{
+	ngrokv1alpha1.KubernetesOperatorFeatureBindings: "Bindings",
+	ngrokv1alpha1.KubernetesOperatorFeatureIngress:  "Ingress",
+	ngrokv1alpha1.KubernetesOperatorFeatureGateway:  "Gateway",
+}
+
+// KubernetesOperatorReconciler reconciles a KubernetesOperator object
+type KubernetesOperatorReconciler struct {
+	client.Client
+	Scheme     *runtime.Scheme
+	controller *controller.BaseController[*ngrokv1alpha1.KubernetesOperator]
+
+	Log            logr.Logger
+	Recorder       record.EventRecorder
+	NgrokClientset ngrokapi.Clientset
+
+	// Namespace where the BindingConfiguration lives
+	Namespace string
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KubernetesOperatorReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	if r.NgrokClientset == nil {
+		return errors.New("NgrokClientset is required")
+	}
+
+	r.controller = &controller.BaseController[*ngrokv1alpha1.KubernetesOperator]{
+		Kube:     r.Client,
+		Log:      r.Log,
+		Recorder: r.Recorder,
+
+		Namespace: &r.Namespace,
+
+		StatusID: func(obj *ngrokv1alpha1.KubernetesOperator) string { return obj.Status.ID },
+		Create:   r.create,
+		Update:   r.update,
+		Delete:   r.delete,
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ngrokv1alpha1.KubernetesOperator{}).
+		WithEventFilter(
+			predicate.Or(
+				predicate.AnnotationChangedPredicate{},
+				predicate.GenerationChangedPredicate{},
+			),
+		).
+		Complete(r)
+}
+
+// +kubebuilder:rbac:groups=ngrok.k8s.ngrok.com,resources=kubernetesoperators,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=ngrok.k8s.ngrok.com,resources=kubernetesoperators/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=ngrok.k8s.ngrok.com,resources=kubernetesoperators/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the BindingConfiguration object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/reconcile
+func (r *KubernetesOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	return r.controller.Reconcile(ctx, req, new(ngrokv1alpha1.KubernetesOperator))
+}
+
+func (r *KubernetesOperatorReconciler) create(ctx context.Context, ko *ngrokv1alpha1.KubernetesOperator) error {
```

Comment:
I see that the API registration is happening in the reconcile loop rather then on startup. I think this changes the expected behaviour of the ngrok-op where previously we would not want the container to ready up and we do not want bindingconfig to be reconciled unless the ngrok-op can claim/register with the API first.